checks.d/mysql.py

```diff
@@ -141,96 +357,193 @@ def _connect(self, host, port, mysql_sock, user, password, defaults_file):
+                    if slave_sql_running:
+                        slave_sql_running = (slave_sql_running.lower.strip() == "yes")
+
+                    if not (slave_io_running is None and slave_sql_running is None):
+                        if slave_io_running and slave_sql_running:
+                            slave_running = 1
+                            slave_running_status = AgentCheck.OK
+                        elif not slave_io_running and not slave_sql_running:
+                            slave_running = 0
+                            slave_running_status = AgentCheck.CRITICAL
+                        else:
+                            # not everything is running smoothly
+                            slave_running = 1
+                            slave_running_status = AgentCheck.WARNING
+
+            self.service_check(self.SLAVE_SERVICE_CHECK_NAME, slave_running_status, tags=tags)
+
+
+        self._rate_or_gauge_vars(metrics, results, tags)
 
         # Collect custom query metrics
         # Max of 20 queries allowed
         if isinstance(queries, list):
             for index, check in enumerate(queries[:self.MAX_CUSTOM_QUERIES]):
-                self._collect_dict(check['type'], {check['field']: check['metric']}, check['query'], db, tags=tags)
+                self._collect_dict(check['type'], {check['field']: check[
+                                   'metric']}, check['query'], db, tags=tags)
 
             if len(queries) > self.MAX_CUSTOM_QUERIES:
                 self.warning("Maximum number (%s) of custom queries reached.  Skipping the rest."
                              % self.MAX_CUSTOM_QUERIES)
 
-
     def _collect_metadata(self, db, host):
         self._get_version(db, host)
 
-    def _rate_or_gauge_statuses(self, statuses, dbResults, tags):
-        for status, metric in statuses.iteritems():
+    def _rate_or_gauge_vars(self, variables, dbResults, tags):
+        for variable, metric in variables.iteritems():
             metric_name, metric_type = metric
-            value = self._collect_scalar(status, dbResults)
-            if value is not None:
-                if metric_type == RATE:
-                    self.rate(metric_name, value, tags=tags)
-                elif metric_type == GAUGE:
-                    self.gauge(metric_name, value, tags=tags)
-
-    def _version_greater_502(self, db, host):
-        # show global status was introduced in 5.0.2
+            for tag, value in self._collect_all_scalars(variable, dbResults):
+                metric_tags = list(tags)
+                if tag:
+                    metric_tags.append(tag)
+                if value is not None:
+                    if metric_type == RATE:
+                        self.rate(metric_name, value, tags=metric_tags)
+                    elif metric_type == GAUGE:
+                        self.gauge(metric_name, value, tags=metric_tags)
+
+    def _version_compatible(self, db, host, compat_version):
```

Comment:
This can be rewritten more concisely in:
```python
    def _version_compatible(self, db, host, compat_version):
        # some patch version numbers contain letters (e.g. 5.0.51a)
        # so let's be careful when we compute the version number
        try:
            mysql_version = self._get_version(db, host)
        except Exception, e:
            self.warning("Cannot compute mysql version, assuming it's older.: %s"
                         % str(e))
            return False
        self.log.debug("MySQL version %s" % mysql_version)

        patchlevel = int(re.match(r"([0-9]+)", mysql_version[2]).group(1))
        version = (int(mysql_version[0]), int(mysql_version[1]), patchlevel)

        return version > compat_version
```